File: backend/KtorApp/src/main/kotlin/com/example/routes/TextRoutes.kt
package com.example.routes

import io.ktor.server.routing.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.http.*
import com.example.utils.DatabaseFactory
import com.example.utils.BackgroundProcessing
import com.example.utils.toBook
import com.example.utils.toTextBlock
import com.mongodb.client.model.Updates
import org.bson.Document

fun Route.textRoutes() {

    // Маршрут для получения текста в зависимости от текущего режима
    post("/get_text") {
        val parameters = call.receiveParameters()
        val idUser = parameters["id_user"]
        val idBook = parameters["id_book"]

        if (idUser == null || idBook == null) {
            call.respond(HttpStatusCode.BadRequest, "ID пользователя или книги не предоставлен")
            return@post
        }

        val booksCollection = DatabaseFactory.getBooksCollection()
        val bookDoc = booksCollection.find(
            Document("_id", idBook).append("userId", idUser)
        ).firstOrNull()

        if (bookDoc == null) {
            call.respond(HttpStatusCode.NotFound, "Книга не найдена или не принадлежит пользователю")
            return@post
        }

        val book = bookDoc.toBook()
        val currentBlockIndex = book.blockStopBook
        val idBlock = book.textBlockIds.getOrNull(currentBlockIndex)

        if (idBlock == null) {
            call.respond(HttpStatusCode.NotFound, "Текстовый блок не найден")
            return@post
        }

        val textBlocksCollection = DatabaseFactory.getTextBlocksCollection()
        val textBlockDoc = textBlocksCollection.find(Document("_id", idBlock)).firstOrNull()

        if (textBlockDoc == null) {
            call.respond(HttpStatusCode.NotFound, "Текстовый блок не найден")
            return@post
        }

        val textBlock = textBlockDoc.toTextBlock()

        when (book.mode) {
            "summarization_time" -> {
                // Если суммированный текст с учетом времени не готов, запускаем фоновую обработку
                if (textBlock.summaryTime == null) {
                    // Запускаем фоновую обработку
                    BackgroundProcessing.processSummarizationForBlock(textBlock, mode = "summarization_time")
                    call.respond(HttpStatusCode.Accepted, "Суммирование с учетом времени чтения запущено. Попробуйте позже.")
                } else {
                    call.respond(mapOf("text" to textBlock.summaryTime, "id_block" to idBlock))
                }
            }
            "summarization" -> {
                // Если суммированный текст не готов, запускаем фоновую обработку
                if (textBlock.summary == null) {
                    // Запускаем фоновую обработку
                    BackgroundProcessing.processSummarizationForBlock(textBlock, mode = "summarization")
                    call.respond(HttpStatusCode.Accepted, "Суммирование запущено. Попробуйте позже.")
                } else {
                    call.respond(mapOf("text" to textBlock.summary, "id_block" to idBlock))
                }
            }
            "questions_original_text" -> {
                call.respond(mapOf("text" to textBlock.original, "id_block" to idBlock))
            }
            else -> {
                call.respond(HttpStatusCode.BadRequest, "Неверный режим книги")
            }
        }
    }

    // Маршрут для получения вопросов по текстовому блоку
    post("/get_questions") {
        val parameters = call.receiveParameters()
        val idUser = parameters["id_user"]
        val idBook = parameters["id_book"]
        val idBlock = parameters["id_block"]

        if (idUser == null || idBook == null || idBlock == null) {
            call.respond(HttpStatusCode.BadRequest, "Не предоставлены необходимые параметры")
            return@post
        }

        val booksCollection = DatabaseFactory.getBooksCollection()
        val bookDoc = booksCollection.find(
            Document("_id", idBook).append("userId", idUser)
        ).firstOrNull()

        if (bookDoc == null) {
            call.respond(HttpStatusCode.NotFound, "Книга не найдена или не принадлежит пользователю")
            return@post
        }

        val textBlocksCollection = DatabaseFactory.getTextBlocksCollection()
        val textBlockDoc = textBlocksCollection.find(Document("_id", idBlock)).firstOrNull()

        if (textBlockDoc == null) {
            call.respond(HttpStatusCode.NotFound, "Текстовый блок не найден")
            return@post
        }

        val textBlock = textBlockDoc.toTextBlock()

        if (textBlock.questions.isEmpty()) {
            // Генерируем вопросы и ответы
            val (questions, answers) = BackgroundProcessing.generateQuestionsForBlock(textBlock.original)
            // Обновляем текстовый блок в базе данных
            textBlocksCollection.updateOne(
                Document("_id", idBlock),
                Updates.combine(
                    Updates.set("questions", questions),
                    Updates.set("rightAnswers", answers)
                )
            )
            call.respond(mapOf("questions" to questions, "answers" to answers))
        } else {
            call.respond(mapOf("questions" to textBlock.questions, "answers" to textBlock.rightAnswers))
        }
    }

    // Маршрут для перехода к следующему блоку текста
    post("/next_block_text") {
        val parameters = call.receiveParameters()
        val idUser = parameters["id_user"]
        val idBook = parameters["id_book"]

        if (idUser == null || idBook == null) {
            call.respond(HttpStatusCode.BadRequest, "ID пользователя или книги не предоставлен")
            return@post
        }

        val booksCollection = DatabaseFactory.getBooksCollection()
        val bookDoc = booksCollection.find(
            Document("_id", idBook).append("userId", idUser)
        ).firstOrNull()

        if (bookDoc == null) {
            call.respond(HttpStatusCode.NotFound, "Книга не найдена или не принадлежит пользователю")
            return@post
        }

        val book = bookDoc.toBook()
        val nextBlockIndex = book.blockStopBook + 1
        if (nextBlockIndex >= book.textBlockIds.size) {
            call.respond(HttpStatusCode.BadRequest, "Это последний блок текста")
            return@post
        }

        booksCollection.updateOne(
            Document("_id", idBook),
            Updates.set("blockStopBook", nextBlockIndex)
        )
        call.respond(mapOf("message" to "Переход к следующему блоку выполнен"))
    }

    // Маршрут для перехода к предыдущему блоку текста
    post("/back_block_text") {
        val parameters = call.receiveParameters()
        val idUser = parameters["id_user"]
        val idBook = parameters["id_book"]

        if (idUser == null || idBook == null) {
            call.respond(HttpStatusCode.BadRequest, "ID пользователя или книги не предоставлен")
            return@post
        }

        val booksCollection = DatabaseFactory.getBooksCollection()
        val bookDoc = booksCollection.find(
            Document("_id", idBook).append("userId", idUser)
        ).firstOrNull()

        if (bookDoc == null) {
            call.respond(HttpStatusCode.NotFound, "Книга не найдена или не принадлежит пользователю")
            return@post
        }

        val book = bookDoc.toBook()
        val previousBlockIndex = book.blockStopBook - 1
        if (previousBlockIndex < 0) {
            call.respond(HttpStatusCode.BadRequest, "Это первый блок текста")
            return@post
        }

        booksCollection.updateOne(
            Document("_id", idBook),
            Updates.set("blockStopBook", previousBlockIndex)
        )
        call.respond(mapOf("message" to "Переход к предыдущему блоку выполнен"))
    }
}

File: backend/routes/TextRoutes.py
from flask import request, jsonify
from fastapi import APIRouter
from pymongo import MongoClient
from bson import ObjectId
from utils.BackgroundProcessing import BackgroundProcessing

text_router = APIRouter()
client = MongoClient('mongodb://localhost:27017/')
db = client['your_database_name']

@text_router.route('/get_text', methods=['POST'])
def get_text():
    id_user = request.form.get('id_user')
    id_book = request.form.get('id_book')

    if not id_user or not id_book:
        return jsonify({"error": "ID пользователя или книги не предоставлен"}), 400

    books_collection = db.books
    book_doc = books_collection.find_one({"_id": ObjectId(id_book), "userId": id_user})

    if not book_doc:
        return jsonify({"error": "Книга не найдена или не принадлежит пользователю"}), 404

    current_block_index = book_doc['blockStopBook']
    id_block = book_doc['textBlockIds'][current_block_index] if current_block_index < len(book_doc['textBlockIds']) else None

    if not id_block:
        return jsonify({"error": "Текстовый блок не найден"}), 404

    text_blocks_collection = db.text_blocks
    text_block_doc = text_blocks_collection.find_one({"_id": ObjectId(id_block)})

    if not text_block_doc:
        return jsonify({"error": "Текстовый блок не найден"}), 404

    mode = book_doc['mode']

    if mode == "summarization_time":
        if text_block_doc.get('summaryTime') is None:
            BackgroundProcessing.process_summarization_for_block(text_block_doc, mode="summarization_time")
            return jsonify({"message": "Суммирование с учетом времени чтения запущено. Попробуйте позже."}), 202
        else:
            return jsonify({"text": text_block_doc['summaryTime'], "id_block": str(id_block)})
    elif mode == "summarization":
        if text_block_doc.get('summary') is None:
            BackgroundProcessing.process_summarization_for_block(text_block_doc, mode="summarization")
            return jsonify({"message": "Суммирование запущено. Попробуйте позже."}), 202
        else:
            return jsonify({"text": text_block_doc['summary'], "id_block": str(id_block)})
    elif mode == "questions_original_text":
        return jsonify({"text": text_block_doc['original'], "id_block": str(id_block)})
    else:
        return jsonify({"error": "Неверный режим книги"}), 400

@text_router.route('/get_questions', methods=['POST'])
def get_questions():
    id_user = request.form.get('id_user')
    id_book = request.form.get('id_book')
    id_block = request.form.get('id_block')

    if not id_user or not id_book or not id_block:
        return jsonify({"error": "Не предоставлены необходимые параметры"}), 400

    books_collection = db.books
    book_doc = books_collection.find_one({"_id": ObjectId(id_book), "userId": id_user})

    if not book_doc:
        return jsonify({"error": "Книга не найдена или не принадлежит пользователю"}), 404

    text_blocks_collection = db.text_blocks
    text_block_doc = text_blocks_collection.find_one({"_id": ObjectId(id_block)})

    if not text_block_doc:
        return jsonify({"error": "Текстовый блок не найден"}), 404

    if not text_block_doc.get('questions'):
        questions, answers = BackgroundProcessing.generate_questions_for_block(text_block_doc['original'])
        text_blocks_collection.update_one(
            {"_id": ObjectId(id_block)},
            {"$set": {"questions": questions, "rightAnswers": answers}}
        )
        return jsonify({"questions": questions, "answers": answers})
    else:
        return jsonify({"questions": text_block_doc['questions'], "answers": text_block_doc['rightAnswers']})

@text_router.route('/next_block_text', methods=['POST'])
def next_block_text():
    id_user = request.form.get('id_user')
    id_book = request.form.get('id_book')

    if not id_user or not id_book:
        return jsonify({"error": "ID пользователя или книги не предоставлен"}), 400

    books_collection = db.books
    book_doc = books_collection.find_one({"_id": ObjectId(id_book), "userId": id_user})

    if not book_doc:
        return jsonify({"error": "Книга не найдена или не принадлежит пользователю"}), 404

    next_block_index = book_doc['blockStopBook'] + 1
    if next_block_index >= len(book_doc['textBlockIds']):
        return jsonify({"error": "Это последний блок текста"}), 400

    books_collection.update_one(
        {"_id": ObjectId(id_book)},
        {"$set": {"blockStopBook": next_block_index}}
    )
    return jsonify({"message": "Переход к следующему блоку выполнен"})

@text_router.route('/back_block_text', methods=['POST'])
def back_block_text():
    id_user = request.form.get('id_user')
    id_book = request.form.get('id_book')

    if not id_user or not id_book:
        return jsonify({"error": "ID пользователя или книги не предоставлен"}), 400

    books_collection = db.books
    book_doc = books_collection.find_one({"_id": ObjectId(id_book), "userId": id_user})

    if not book_doc:
        return jsonify({"error": "Книга не найдена или не принадлежит пользователю"}), 404

    previous_block_index = book_doc['blockStopBook'] - 1
    if previous_block_index < 0:
        return jsonify({"error": "Это первый блок текста"}), 400

    books_collection.update_one(
        {"_id": ObjectId(id_book)},
        {"$set": {"blockStopBook": previous_block_index}}
    )
    return jsonify({"message": "Переход к предыдущему блоку выполнен"})

if __name__ == '__main__':
    text_router.run(debug=True)

